src/node_crypto.cc

```diff
@@ -4349,6 +4351,123 @@ void Hash::HashDigest(const FunctionCallbackInfo<Value>& args) {
 }
 
 
+void Mac::Initialize(Environment* env, Local<Object> target) {
+  Local<FunctionTemplate> t = env->NewFunctionTemplate(New);
+  t->InstanceTemplate()->SetInternalFieldCount(Mac::kInternalFieldCount);
+
+  env->SetProtoMethod(t, "update", Update);
+  env->SetProtoMethod(t, "final", Final);
+
+  target->Set(env->context(),
+              FIXED_ONE_BYTE_STRING(env->isolate(), "Mac"),
+              t->GetFunction(env->context()).ToLocalChecked()).Check();
+}
+
+
+Mac::Mac(Environment* env, Local<Object> wrap, int nid, EVPMDPointer&& mdctx)
+    : BaseObject(env, wrap)
+    , nid_(nid)
+    , mdctx_(std::move(mdctx)) {
+  MakeWeak();
+}
+
+
+void Mac::New(const FunctionCallbackInfo<Value>& args) {
+  MarkPopErrorOnReturn mark_pop_error_on_return;
+  Environment* env = Environment::GetCurrent(args);
+
+  CHECK(args[0]->IsInt32());
+  const int nid = args[0].As<Int32>()->Value();
+
+  CHECK(args[1]->IsArrayBufferView());
+  ByteSource key = ByteSource::FromBuffer(args[1]);
+
+  const EVP_MD* md = nullptr;
+  if (nid == EVP_PKEY_HMAC) {
+    CHECK(args[2]->IsString());
+    const node::Utf8Value name(env->isolate(), args[2]);
+    md = EVP_get_digestbyname(*name);
+    if (md == nullptr)
+      return CheckThrow(env, SignBase::Error::kSignUnknownDigest);
+  }
+
+  EVPKeyPointer key_ptr;
+  if (nid == EVP_PKEY_CMAC) {
+    CHECK(args[2]->IsString());
+    const node::Utf8Value name(env->isolate(), args[2]);
+    const EVP_CIPHER* cipher = EVP_get_cipherbyname(*name);
+    if (cipher == nullptr) return env->ThrowError("Unknown cipher");
+    key_ptr.reset(
+        EVP_PKEY_new_CMAC_key(nullptr,
+                              reinterpret_cast<const unsigned char*>(key.get()),
+                              key.size(), cipher));
+  } else {
+    key_ptr.reset(
+        EVP_PKEY_new_raw_private_key(
+            nid, nullptr, reinterpret_cast<const unsigned char*>(key.get()),
+            key.size()));
+  }
+
+  ManagedEVPPKey pkey(std::move(key_ptr));
+  EVPMDPointer mdctx(EVP_MD_CTX_new());
+
+
+  EVP_PKEY_CTX* pkctx = nullptr;
+  if (!EVP_DigestSignInit(mdctx.get(), &pkctx, md, nullptr, pkey.get()))
+    return ThrowCryptoError(env, ERR_get_error(), "EVP_DigestSignInit");
+
+  // TODO(bnoordhuis) Call EVP_PKEY_CTX_ctrl() or EVP_PKEY_CTX_ctrl_str().
```

Comment:
nit... can you include a quick explanation why this is needed in case someone else wants to come in later and address the todo